Add gitea support

pachiras opened the pull request:
Hello, I've changed three files to support gitea server(https://gitea.io/en-us/)

modules/nextflow/src/main/groovy/nextflow/scm/RepositoryProvider.groovy
modules/nextflow/src/main/groovy/nextflow/scm/ProviderConfig.groovy
modules/nextflow/src/main/groovy/nextflow/scm/GiteaRepositoryProvider.groovy

As for other files, I'm not sure why the versions were changed.

I'm not familiar with this github pull system.
If my operation was not appropriate, please tell me.

pditommaso commented on docs/conf.py:
```diff
@@ -48,9 +48,9 @@
 # The full version, including alpha/beta/rc tags.
-release = '18.12.0-edge'
+release = '19.1.0-SNAPSHOT'
 
```
These numbers are automatically updates, do not add this file to the commit.

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/Const.groovy:
```diff
@@ -50,17 +50,17 @@ class Const {
-    static public final int APP_BUILDNUM = 5048
+    static public final int APP_BUILDNUM = 5050
 
 
```
As before, These numbers are automatically updates, do not add this file to the commit.

pditommaso commented on modules/nextflow/src/main/groovy/nextflow/scm/GiteaRepositoryProvider.groovy:
```diff
@@ -0,0 +1,77 @@
+ * Implements a repository provider for Gitea service
+ *
+ * @author Paolo Di Tommaso <paolo.ditommaso@gmail.com>
+ */
```
Add you name to keep track of the authorship.

pditommaso commented on modules/nextflow/src/test/groovy/nextflow/scm/ProviderConfigTest.groovy:
```diff
@@ -167,7 +167,8 @@ class ProviderConfigTest extends Specification {
         then:
-        result.size() == 5
+//        result.size() == 5
+        result.size() == 6
```
Ok, just remove the commented line.

pditommaso commented on nextflow:
```diff
@@ -15,7 +15,7 @@
 [[ "$NXF_DEBUG" == 'x' ]] && set -x
-NXF_VER=${NXF_VER:-'18.12.0-edge'}
+NXF_VER=${NXF_VER:-'19.1.0-SNAPSHOT'}
 NXF_ORG=${NXF_ORG:-'nextflow-io'}
```
As before, these numbers are automatically updates, do not add this file to the commit.

pditommaso commented on modules/nextflow/work/39/dc39c45cea68701a560cd38b10ce42/.command.out:
```diff
@@ -0,0 +1 @@
+ciao
```
None of the files under the `work` path should be commit.

pditommaso commented:
Hi Akira, maybe there's a misunderstanding. I meant to revert the changes in that files, not to delete the files. You should ab able to do that (soft) resetting your local commit, discard that changes, then commit and (force) pushing again. 

Also please include a section how to configure Gitea in the documentation https://github.com/nextflow-io/nextflow/blob/master/docs/sharing.rst#gitlab-credentials

pachiras commented:
Hi Paolo,

I've soft-reset my local repository back to your last commit and git-added my files. Now, I think, all the differences are my modified codes only.

I also mentioned about "gitea" in the 'Private server configuration' section in `docs/sharing.rst`.

pditommaso commented:
Excellent, but a bit more documentation would be required. Could you add a section after [GitLab credentials](https://www.nextflow.io/docs/latest/sharing.html#gitlab-credentials) showing how to setup Gitea? It should describe what are the minimal required information eg. user name? password? is there any access token? how to get it? Think for a user who never used Gitea. Thanks!

pachiras commented:
OK, I've updated the document for Gitea. But I want you to correct the sentences to make them more natural and readable since I'm not a good English writer.

Furthermore, I've modified `nextflow/modules/nextflow/src/main/groovy/nextflow/scm/GitUrl.groovy` to support subdirectory in URL, such as `http://your-domain.org/gitea/` where the repository service is provided under `gitea/`. 

Thanks.

pditommaso commented:
> OK, I've updated the document for Gitea

Nice!

> Furthermore, I've modified nextflow/modules/nextflow/src/main/groovy/nextflow/scm/GitUrl.groovy to support subdirectory in URL.

Not sure to understand this need. Could you please add the relative tests in the [GitUrlTest](https://github.com/nextflow-io/nextflow/blob/b49e6d74dae782def6be2b087d06fdd8fa6b9291/modules/nextflow/src/test/groovy/nextflow/scm/GitUrlTest.groovy#L29-L54) class.

pachiras commented:
> Not sure to understand this need.
I hope those people who works in a small private network, including me, think it useful.
Anyway, I added some tests in GitUrlTest.groovy.

pditommaso commented:
Thanks for adding the tests. Now I understand your need but not sure to agree with stripping the first path of the URL. Is it strictly necessary for your use case  ?

pachiras commented:
Thank you for your comment, Paolo.

In my case I have a single global IP hosting many services as explained in this link:
https://gist.github.com/soheilhy/8b94347ff8336d971ad0

Each subfolder is mapped to a port on the same host. And those ports are assigned to different applications. I think this is a popular way to share an IP address for multiple services.

pditommaso commented:
Let me think a bit if there's an alternative to make that logic configurable.

pachiras commented:
Sure. If you think this functionality will harm the integrity of the software, I will not insist.

pditommaso commented:
I think I've a better way to manage sub-paths in the Git URL. The path prefix should go in the [server](https://github.com/nextflow-io/nextflow/blob/001dea4e427911624e34c3d2111eaa3bb29d0038/modules/nextflow/src/test/groovy/nextflow/scm/ProviderConfigTest.groovy#L119) setting. Therefore that info can be used to [remove the path prefix](https://github.com/nextflow-io/nextflow/blob/001dea4e427911624e34c3d2111eaa3bb29d0038/modules/nextflow/src/main/groovy/nextflow/scm/AssetManager.groovy#L295-L297) from the Git URL. 

See an example [here](https://github.com/nextflow-io/nextflow/blob/001dea4e427911624e34c3d2111eaa3bb29d0038/modules/nextflow/src/test/groovy/nextflow/scm/AssetManagerTest.groovy#L491-L494). 

Let me know if this work.

pachiras commented:
Hello Paolo,

Thank you for your time for considering this. I think your solution is more sophisticated than mine. I'm testing the code. Please wait.

pachiras commented:
Sorry to be late, but here is the test result.

Firstly I removed `endpoint` entry from my `$HOME/.nextflow/scm`.

```shell
providers {
	gitea {
		server = 'https://mydomain.com'
		platform = 'gitea'
        	user = 'pachiras'
        	password = 'mypassword'
		token = '65eaa9c8ef52460d22a93307fe0aee76289dc675'
	}
}
```

Then I typed some commands:
```shell
[pachiras@ngs1 ~]$ mkdir workspace
[pachiras@ngs1 ~]$ cd workspace/
[pachiras@ngs1 workspace]$ git clone -b pachiras-gitea_devel https://github.com/nextflow-io/nextflow.git
Cloning into 'nextflow'...
remote: Enumerating objects: 377, done.
remote: Counting objects: 100% (377/377), done.
remote: Compressing objects: 100% (245/245), done.
remote: Total 44434 (delta 134), reused 302 (delta 94), pack-reused 44057
Receiving objects: 100% (44434/44434), 32.78 MiB | 1006.00 KiB/s, done.
Resolving deltas: 100% (26245/26245), done.
[pachiras@ngs1 workspace]$ cd nextflow/
[pachiras@ngs1 nextflow]$ git rev-parse HEAD
001dea4e427911624e34c3d2111eaa3bb29d0038
[pachiras@ngs1 nextflow]$ git branch
* pachiras-gitea_devel
[pachiras@ngs1 nextflow]$ make pack
./gradlew packAll
Starting a Gradle Daemon (subsequent builds will be faster)
> Task :nextflow:compileGroovy
> Task :packAll
BUILD SUCCESSFUL in 42s
15 actionable tasks: 15 executed
[pachiras@ngs1 nextflow]$ chmod u+x build/releases/nextflow-19.02.0-SNAPSHOT-all
[pachiras@ngs1 nextflow]$ build/releases/nextflow-19.02.0-SNAPSHOT-all run https://mydomain.com/gitea/NextFlow/hello.git
N E X T F L O W  ~  version 19.01.0-edge
Pulling gitea/NextFlow/hello ...
Cannot find `gitea/NextFlow/hello` -- Make sure exists a Gitea repository at this address `https://mydomain.com/gitea/NextFlow/hello`
```

Well, I think we need to remove 'gitea' from the project name at somewhere maybe in GitUrl.groovy.

pditommaso commented:
> Firstly I removed endpoint entry

It's not required only the `endpoint` (API) is supposed to be different from `server` path. 

I think in your case `server` should be `https://mydomain.com/gitea` not `https://mydomain.com`. 


> Well, I think we need to remove 'gitea' from the project name at somewhere maybe in GitUrl.groovy.

What do you mean?

pditommaso commented:
Nice. I'm going to merge it.

pditommaso commented:
Thanks for your contribution. I've merged it manually.
